features/FEATURE_BLE/targets/TARGET_CORDIO/BleImplementationForward.h

```diff
@@ -0,0 +1,164 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2019 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef MBED_OS_EXAMPLE_BLINKY_BLEIMPLEMENTATIONFORWARD_H
+#define MBED_OS_EXAMPLE_BLINKY_BLEIMPLEMENTATIONFORWARD_H
+
+////////////////////////////////////////////////////////////////////////////////
+// Forward declarations of the implementation types
+//
+namespace ble {
+
+    namespace interface {
+
+        template<class Impl>
+        class LegacyGap;
+
+        template<class Impl>
+        class Gap;
+
+        template<class Impl>
+        class GattClient;
+
+        template<class Impl>
+        class SecurityManager;
+
+        template<class Impl>
+        class GattServer;
+
+    } // namespace interface
+
+
+    namespace generic {
+
+        template<
+            template<class> class TGapImpl,
+            class PalSecurityManager,
+            class ConnectionEventMonitorEventHandler
+        >
+        class GenericGap;
+
+        template <
+            template<class> class TPalSecurityManager,
+            template<class> class SigningMonitor
+        >
+        class GenericSecurityManager;
+
+        template<class PalGattClient, class SigningMonitorEventHandler>
+        class GenericGattClient;
+
+    } // namespace generic
+
+    namespace pal {
+
+        template<class Impl, class EventHandler>
+        class SecurityManager;
+
+        template<class AttClient>
+        class AttClientToGattClientAdapter;
+
+        namespace vendor {
+            namespace cordio {
+
+                template <class EventHandler>
+                class CordioSecurityManager;
+
+                template<class EventHandler>
+                class Gap;
+
+                class CordioAttClient;
+            } // cordio
+        } // vendor
+    } // pal
+
+    namespace vendor {
+        namespace cordio {
+
+            template<class EventHandler>
+            class SigningEventMonitor;
+
+            class GattServer;
+
+        } // cordio
+    } // vendor
+} // ble
+
+
+// implementation assembly
+namespace ble {
+    namespace impl {
+        // SECURITY MANAGER
+        typedef generic::GenericSecurityManager<
+            pal::vendor::cordio::CordioSecurityManager,
+            vendor::cordio::SigningEventMonitor
+        > GenericSecurityManagerImpl;
```

Comment:
why both "impl" and "Impl" suffix

Reply:
I've added some information in the description; basically there is implementation classes and implementation interfaces. In that case `GenericSecurityManagerImpl` is the implementation class. The interface imported in the global namespace will be `ble::interface::SecurityManager<GenericSecurityManagerImpl>`. We don't want to import `GenericSecurityManagerImpl` as it requires visibility of `GenericSecurityManagerImpl` and may export thing differently than `ble::interface::SecurityManager` (visibility of function can change as an example).